src/Shortcode.php

```diff
@@ -177,10 +182,37 @@ private function embedTicket($id): string
 
     private function embedUser(?string $username): string
     {
+        if (empty($username)) {
+            return '';
+        }
+
         return GetUserAndTooltipDiv($username, false);
     }
 
-    private function linkifyBasicURLs(string $text): string
+    private function autolinkRetroachievementsUrls(string $text): string
+    {
+        // see https://stackoverflow.com/a/2271552/580651:
+        // [...] it's probably safe to assume a semicolon at the end of a URL is meant as sentence punctuation.
+        // The same goes for other sentence-punctuation characters like periods, question marks, quotes, etc..
+        // lookahead: (?<![!,.?;:"\'()-])
+        return (string) preg_replace(
+            '~
+                (?:https?://)?      # Optional scheme. Either http or https.
+                (?:www.)?           # Optional subdomain.
+                retroachievements\.        # Host.
+                ([\w!#$%&\'()*+,./:;=?@\[\]-]+
+                (?<![!,.?;:"\'()]))
+                (?!                 # Assert URL is not pre-linked.
+                  [^<>]*>           # Either inside a start tag,
+                  | [^<>]*</a>      # End recognized pre-linked alts.
+                )                   # End negative lookahead assertion.
+            ~ix',
+            '<a href="https://retroachievements.$1">https://retroachievements.$1</a>',
```

Comment:
My regex-fu is not very strong, but wouldn't `$1` be the `https` or `www` capture?

Reply:
Same here. Yet, I learned a while ago that `(?` are non-capturing groups; meaning they will be treated as an entity but not exposed as a match. I trusted regex101, stackoverflow and the tests in this case tbh.
The reason for not including `org` in the domain stems from pretty much the same vein of not having a damn clue what regex wants from me. The sentence-punctuation characters clause stopped working as soon as i wanted to include the tld in there for some reason. Chose the easy way out. But it works 🤷